vmm/src/default_syscalls/filters.rs

```diff
@@ -12,20 +11,31 @@ use seccomp::{
 // even though they are defined in musl libc:
 // https://git.musl-libc.org/cgit/musl/tree/arch/aarch64/bits/syscall.h.in.
 // Submitted issue in rust-lang: https://github.com/rust-lang/libc/issues/1348.
-const SYS_fcntl: ::std::os::raw::c_long = 25;
-const SYS_lseek: ::std::os::raw::c_long = 62;
-const SYS_newfstatat: ::std::os::raw::c_long = 79;
-const SYS_fstat: ::std::os::raw::c_long = 80;
-const SYS_mmap: ::std::os::raw::c_long = 222;
+#[allow(non_upper_case_globals)]
+#[cfg(target_arch = "aarch64")]
```

Comment:
Well, it's the same with glibc vs musl syscalls. The problem here is that most of them are the same across platforms. I find it weird to duplicate them when only 3 syscalls are different.

Reply:
3 for now. This PR is living proof that Rust version changes bring in syscall tweaks, and these tweaks might differ between `x86_64` and `aarch64`. There are 3 options I can see for composing the full seccomp filter -
1. one file with `$[cfg()]`s. **Pros**: you don't need to look all over the place when you need to tweak the filters. **Cons**: becomes harder and harder to read across architectures.
1. one file per architecture. **Pros**: easier to read. If you have an arch-specific problem, it's easy to narrow down the source. **Cons**: if you don't have an arch-specific problem, you have N (= no of supported arch's) files to modify.
1. one "common" file with stuff that's shared across arch's, and "diff" arch-specific files. **Pros**: keeps arch stuff separated while avoiding code deduplication. **Cons**: larger number of files than the other options, harder to track down an issue.
@firecracker-microvm/compute-capsule what do you think?

Reply:
Just one note: in this PR there is no syscall different before aarch64 and x86. The different syscalls are from before. So I would say that the different syscalls are rather an exception than the usual behavior.

Reply:
I agree with @aghecenco , that putting the syscalls for all architectures in one file makes the code less readable (and as a consequence, more prone to  errors).
Moreover, the syscalls used by a flow of operations are tied to the architecture. The fact that the same syscalls are triggered on both architectures (with only slight differences in their name) is just a coincidence that may change in the future based on the functionality that we may add. 
Take the case where we use a specific aarch64 RTC driver (PL031) for emulating the RTC device. The fact that it does not introduce a different/additional syscall is just a coincidence/happy scenario.
Also, let us take a look at how [musl](https://git.musl-libc.org/cgit/musl/tree/arch/) also puts them in separate files. glibc goes even further and uses a script [make-syscalls](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/make-syscalls.sh;hb=HEAD) that geenrates the syscalls at build time based on [sysdeps/](https://sourceware.org/git/?p=glibc.git;a=tree;f=sysdeps;hb=HEAD){arch}. 
@andreeaflorescu I would be very reticent in saying "there is no syscall different between aarch64 and x86". You can see in [x86_64 musl](https://git.musl-libc.org/cgit/musl/tree/arch/x86_64/bits/syscall.h.in) and [aarch64 musl](https://git.musl-libc.org/cgit/musl/tree/arch/aarch64/bits/syscall.h.in) that their names may be the same but the numbers behind them are entirely architectural specific:
```
--> on aarch64 
#define __NR_accept 202
--> on x86_64 
#define __NR_accept 43

```

Reply:
I think that, currently, keeping the list in one file is more manageable. Most of the syscalls that we use are common across architectures and libc flavors. Also, I see no evidence to support the argument that this is just a coincidence and that the syscall lists will diverge significantly any time soon; there is way more common code/logic across architectures, than there are differences.

Down the road, however, we will need to rethink how we manage the syscall list. Each thread should have its own seccomp filter, so this list will be keyed by a three-tuple (arch, libc, thread), instead of (arch, libc). In this context, choosing the arch as a special key to split the list into files seems kinda arbitrary.

Reply:
> I agree with @aghecenco , that putting the syscalls for all architectures in one file makes the code less readable (and as a consequence, more prone to errors).

I think I should explain better how did I got to the conclusion of having them in the same list, in the same file.

The first reason is that I had to enable the same syscalls on both architectures:
- SYS_accept4
- FUTEX_CMP_REQUEUE_PRIVATE argument to SYS_futex

Initially I did not understand why the tests failed on arm, but pass on x86_64. Then I noticed that they're actually platform specific (this is totally my miss).

The second reason is that different targets (musl, gnu) share the same list in Firecracker because most syscalls are the same with few exceptions. I thought the same applies for different architectures so I did not see a reason why for targets to apply a different approach than for architectures.

If we do keep them separate, we should also separate them per target to keep it consistent across the code.

> Moreover, the syscalls used by a flow of operations are tied to the architecture. The fact that the same syscalls are triggered on both architectures (with only slight differences in their name) is just a coincidence that may change in the future based on the functionality that we may add.

That is true, I didn't think about this angle. In this case I would still keep a common list and only move the specific syscalls in a per-architecture and per-target list.

> Take the case where we use a specific aarch64 RTC driver (PL031) for emulating the RTC device. The fact that it does not introduce a different/additional syscall is just a coincidence/happy scenario.
> Also, let us take a look at how [musl](https://git.musl-libc.org/cgit/musl/tree/arch/) also puts them in separate files. glibc goes even further and uses a script [make-syscalls](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/make-syscalls.sh;hb=HEAD) that geenrates the syscalls at build time based on [sysdeps/](https://sourceware.org/git/?p=glibc.git;a=tree;f=sysdeps;hb=HEAD){arch}.

Only the numbers are different on different architectures. Because the numbers are so different depending on architecture (32 vs 64) and platform (arm, misp, x86) it makes sense that libc and musl places them in different files. This is not something that users of either glibc or musl should care about. That's why syscalls are called the same on all platforms. The way I see it is that libc and musl are adding this abstraction so that developers can write platform-independent code where possible.

> @andreeaflorescu I would be very reticent in saying "there is no syscall different between aarch64 and x86". You can see in [x86_64 musl](https://git.musl-libc.org/cgit/musl/tree/arch/x86_64/bits/syscall.h.in) and [aarch64 musl](https://git.musl-libc.org/cgit/musl/tree/arch/aarch64/bits/syscall.h.in) that their names may be the same but the numbers behind them are entirely architectural specific:
> 
> ```
> --> on aarch64 
> #define __NR_accept 202
> --> on x86_64 
> #define __NR_accept 43
> ```

I think there might be some confusion here. In the kernel the mapping between a syscall define and the implementation of the syscall is made between syscall name (__NR__accept) and implementation and not syscall number (which is different on different platforms/architectures/targets like 202 and 43 in your example) and implementation. We should not care what numbers the syscall defines resolve to. In other words, even though sycalls have different numbers on different platforms they might still have the same implementation.

The best explanation about syscall tables is probably in the documentation about adding a new [syscall to the kernel](https://elixir.bootlin.com/linux/latest/source/Documentation/process/adding-syscalls.rst), specifically the "Generic System Call Implementation", "Compatibility System Calls (Generic)" and "Other Details" sections.

Reply:
I don't have very strong feelings regarding the number of files we use here; I feel like Dan's opinion aligns best with my perspective right now.

Reply:
leaving aside any technical argument, If we move all the syscalls in the same file, we would end up having 4 types of cfg (aarch64, x86_64, musl and gnu) inside one file.  
 @dhrgit Indeed, "keeping a list in one file is more manageable", but keeping a list that is cfg'ed with 4 different labels (4 for the moment) is a mistake imo (at least from an aesthetic view).

Reply:
[Crosvm's sycall defines](https://chromium.googlesource.com/chromiumos/platform/crosvm/+/refs/heads/master/syscall_defines/src/)

Reply:
@dianpopa those are the defines (aka the syscall numbers depending on platform) and not the filters. From their readme: "Lists of syscall numbers in each architecture used to make syscalls not supported in `libc`.". What is the relevance to the whitelist? Crosvm handles seccomp in a very different way. They have policies per architecture/per device.

Reply:
@andreeaflorescu  the relevance lies in the fact that they did not have any problem with duplicating code even though for same arch (i.e x86 and x86_64) they could have very well worked around a lot of duplicated code with cfgs. Anyways, I will summarize my opinion on this. I think that having a list of syscalls names that gets cfg'ed with 2*`no of supported archs` = 4 (for the moment) makes the code less readable, vulnerable to future growth and less suggestive of the functionality it tries to accomplish. As a consequence, I cannot approve of this.